### PR TITLE
SubscriptionSetImpl in JS and PNFuture.remember()

### PR DESCRIPTION
**Pull request by wkal-pubnub**

### CODE REVIEW

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt [hunk 1]
```diff
@@ -7,44 +7,73 @@ import com.pubnub.api.v2.callbacks.Consumer
...
+import kotlinx.atomicfu.locks.ReentrantLock
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
 import kotlin.jvm.JvmName
 
-fun <T> PubNubException.asFuture(): PNFuture<T> = object : PNFuture<T> {
-    override fun async(callback: Consumer<Result<T>>) {
-        callback.accept(Result.failure(this@asFuture))
+private class CompletablePNFuture<T> : PNFuture<T> {
+    private val reentrantLock: ReentrantLock = reentrantLock()
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:16]: what do you mean?
line 11: `import kotlinx.atomicfu.locks.reentrantLock` and it's public

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt [hunk 2]
```diff
@@ -7,44 +7,73 @@ import com.pubnub.api.v2.callbacks.Consumer
...
 import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.locks.ReentrantLock
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
 import kotlin.jvm.JvmName
 
-fun <T> PubNubException.asFuture(): PNFuture<T> = object : PNFuture<T> {
-    override fun async(callback: Consumer<Result<T>>) {
-        callback.accept(Result.failure(this@asFuture))
+private class CompletablePNFuture<T> : PNFuture<T> {
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:15]: line 31: `override fun async` 
are we looking at the same file? :)

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt [hunk 3]
```diff
@@ -7,44 +7,73 @@ import com.pubnub.api.v2.callbacks.Consumer
...
+    private val reentrantLock: ReentrantLock = reentrantLock()
+    private lateinit var result: Result<T>
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:17]: hmm there is no inline class here and it compiles just fine for me... can you see if you have a broken/modified checkout or do a clean build?

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:15]: Sorry. My bad. Looks good.

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:16]: Sorry. My bad. Looks good.

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/kmp/futures.kt:17]: Sorry. My bad. Looks good.

**Review by marcin-cebo — APPROVED** [review]

**Pull request merged**